[FLINK-25769][table] Add internal functions and basic function versioning

twalthr opened the pull request:
## What is the purpose of the change

This adds a basic but sufficient mechanism for function versioning. It adds the concept of internal functions to the stack. Internal functions are fully qualified by name and version. They are not exposed to the user and can thus kept around for a longer period of time. The most recent version of a function is stored in `BuiltInFunctionDefinition` and a new `BuiltInSqlFunction`.

## Brief change log

- Add a flag for internal functions to `BuiltInFunctionDefinition`
- Introduce `BuiltInSqlFunction` and `BuiltInSqlOperator` for easier long-term handling of Calcite functions.
- Adapt some functions in `FlinkSqlOperatorTable` to show the idea and a couple of reference implementations
- Update the first internal function `$REPLICATE_ROWS$1`.


## Verifying this change

This change is already covered by existing tests, such as `CoreModuleTest`.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? JavaDocs

slinkydeveloper commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/module/Module.java:
```diff
@@ -36,17 +36,28 @@
+     *
+     * @return a set of function names
+     */
+    default Set<String> listFunctions(boolean includeInternal) {
```
I'm not a fan of using a boolean to filter something in a `PublicEvolving` interface. Use an enum instead?

slinkydeveloper replied:
In general, I don't understand why you want to expose the internal functions from the `Module`. Module is a public interface, and the change to the interface to list internal functions sounds like it's leaking some internal detail of how our operators behave. Have you considered implementing the resolution of internal functions at another level, e.g. within `FunctionCatalog`?

twalthr replied:
The module system does not the concept of an internal function. Instead it is only able to hide certain functions during listing. I rephrased this part and split it into a separate commit.

slinkydeveloper commented on flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/functions/sql/BuiltInSqlFunction.java:
```diff
@@ -0,0 +1,244 @@
+ * <p>For simple functions, use the provided builder. Otherwise, this class can also be extended.
+ */
+@Internal
+public class BuiltInSqlFunction extends SqlFunction implements BuiltInSqlOperator {
```
Can you move the changes related to this class (and `BuiltInSqlOperator`) in a separate commit, as they seem more of a refactoring hotfix instead?

twalthr replied:
This change is part of the internal/versioning story. However, I will split the refactorings into a separate commit.

slinkydeveloper commented on flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/BuiltInFunctionDefinition.java:
```diff
@@ -90,6 +110,18 @@ public boolean hasRuntimeImplementation() {
+        return isInternal;
+    }
+
+    public String getQualifiedName() {
```
the term "qualified name" IMO is misleading, as it can be easily confused with FQCN from Java terminology. Just identifier?

matriv replied:
+1 for me

matriv replied:
Hm, rethought of that, qualifier, qualified name makes sense for SQL terms.

twalthr commented:
Thanks for the feedback @slinkydeveloper and @matriv. I split the PR into several commits and addressed the module system issue. I also found a bug on the way in the `CoreModule`.

slinkydeveloper commented on flink-table/flink-table-api-java/src/test/java/org/apache/flink/table/module/ModuleManagerTest.java:
```diff
@@ -54,14 +48,15 @@ public void before() {
+        assertThat(manager.getUsedModules())
+                .isEqualTo(Collections.singletonList(CoreModuleFactory.IDENTIFIER));
+        assertThat(manager.getLoadedModules().get(CoreModuleFactory.IDENTIFIER))
+                .isEqualTo(CoreModule.INSTANCE);
```
`isSameAs` to check the instance?

slinkydeveloper commented on flink-table/flink-table-common/src/test/java/org/apache/flink/table/module/CoreModuleTest.java:
```diff
@@ -18,14 +18,56 @@
+                                assertThat(def)
+                                        .isInstanceOfSatisfying(
+                                                BuiltInFunctionDefinition.class,
+                                                builtInDef ->
+                                                        assertThat(builtInDef.getQualifiedName())
+                                                                .isEqualTo("$CAST$1")));
```
You can simplify this with:


```
        assertThat(CoreModule.INSTANCE.getFunctionDefinition("CAST"))
                .hasValueSatisfying(
                        def ->
                                assertThat(def)
                                        .asInstanceOf(InstanceOfAssertFactories.type(BuiltInFunctionDefinition.class))
                                        .extracting(BuiltInFunctionDefinition::getQualifiedName)
                                        .isEqualTo("$CAST$1"))
```

slinkydeveloper reviewed (approved):
Just left a couple of style details

matriv reviewed (approved):
LGTM
